### ISSUE TITLE
how to use iOS Sync/DeltaSync with paginated results?

### ISSUE THREAD

**Issue by airstance** [awslabs/aws-mobile-appsync-sdk-ios#135]
How do I use Sync/DeltaSync with paginated results?

(Posted on the AWS Forums but this is probably a better place to ask the question?)

There's no obvious way to use the iOS library sync/deltasync functionality for anything other than a single response from the service. What if the backend uses something like DynamoDB which may enforce pagination due to several reasons (RCU exhaustion, query size, batchgetitem limit of 100, etc)?

I would attempt to execute a fetch with "nextToken" on the baseQueryHandler but there are a couple of problems. The biggest one is that I don't know if the completion is for runBaseQueryFromCache or runBaseQuery so I don't know what to supply for the cache control parameter. The other problem is that everything will happen at the same time (fetching from cache and source) unless I use a DispatchGroup to enforce synchronous calls.

I'm guessing that I'm just missing something, otherwise this would have been a feature request.

**Comment by airstance** [comment]
... perhaps a return value from baseQueryHandler or some other signal to continue requesting with a new parameter (page/nextToken/whatever).

**Comment by cbommas** [comment]
@airstance 
Thank you for reporting this issue. We will investigate and update this thread when we have more information.

**Comment by palpatim** [comment]
Hi @airstance,

For the first issue:

>  I don't know if the completion is for runBaseQueryFromCache or runBaseQuery 

All `GraphQLResult`s should include a `source` that you can inspect:

```swift
// GraphQLResult<Data> ...
public enum Source {
    case cache
    case server
}

// In calling code result handler:

// Note that the first query may actually have an error because it attempts to decode JSON data with no value.
appSyncClient?.fetch(query: query, cachePolicy: .returnCacheDataDontFetch) { result, _ in
    XCTAssertEqual(result?.source, GraphQLResult.Source.cache)
}

// Even a query against an empty service store will return a `result` with a correctly configured `source`
appSyncClient?.fetch(query: query, cachePolicy: .fetchIgnoringCacheData) { result, _ in
    XCTAssertEqual(result?.source, GraphQLResult.Source.server)
}
```

Thanks for calling this out, I'm adding these conditions to our tests.


For the second issue: @rohandubal other than invoking subsequent queries with new `nextToken`s, is there a recommended pattern for retrieving paginated results?

**Comment by palpatim** [comment]
While we're waiting for @rohandubal to chime in, is this a fair description of the sequence you're concerned about?

- App launches for first time
- App executes `listRecords()` query
- Base query returns `nil` from cache
- Base query returns ResultSet1 from service, with NextToken1 for continuation
- AppSync stores RS1/NT1 as cached result for `listRecords()`
- App executes `listRecords(NT1)` to get next record set
- Base query returns ResultSet2 from service, with NextToken2 for continuation
- AppSync stores RS2/NT2 to local cache as cached result for `listRecords(NT1)`
- App quits & resumes
- Delta sync kicks off, starting with BaseQuery of `listRecords()`
- AppSync retrieves RS1/NT1 from local cache
  - AppSync begins retrieving RS1a/NT1a from service. Meanwhile...
  - ...App executes `listRecords(NT1)` to get next record set
  - AppSync returns cached result for `listRecords(NT1)`
  - Service errs out on invalid `listRecords(NT1)` call since NT1 isn't valid any more
  - ^--- _this is the case we want to avoid_
- AppSync stores RS1a/NT1a as updated cached result for `listRecords()`

If so, I think the immediate workaround would be to use the `source` attribute as described above to ignore next token if the result is coming from the cache, and we can investigate suppressing `nextToken` in cached results as a feature request.

If not, can you help me better understand the case you're trying to solve?

**Comment by javiertc** [comment]
Hi @palpatim, I think that this issue has covered more importance since AppSync has a limit of 1MB for the base query as stated here https://forums.aws.amazon.com/thread.jspa?messageID=857118#857118. Thus, it's really important to paginate in the base query. This is happening to us right now. We used to bring 1500 records in the base query but now we can't do it and we have to paginate.
In the answer above when you say **App executes listRecords(NT1) to get next record set**, what do you mean with it ? do we have to call fetch with the same query that we register with the base query but without using the sync function ?

**Comment by palpatim** [comment]
Hi @javiertc,

> do we have to call fetch with the same query that we register with the base query but without using the sync function

Yes, that's what I mean, but I should have specified: you'd call that from within the base/delta query's result handler, just like you'd do if your app user scrolled to the bottom of the page and you needed to retrieve the next page of results.

Let me attempt to clarify by breaking it down a bit further.

Consider a blog app. In that schema, you might have a `Post` type, some mutations to create, modify, or delete posts, queries to retrieve paginated data, and a subscription that returns updates on *any* of the supported mutations:

```graphql
// Post omitted
// createPost, upvotePost, deletePost mutations omitted

type Query {
  // Note that nextToken is optional. Passing no token returns the first page of results,
  // making it appropriate for the first query.
  listPosts(nextToken: String): [PostConnection]
  listPostsDelta(lastSync: AWSTimestamp, nextToken: String): [PostConnection]
}

type Subscription {
  onDeltaPost: Post
    @aws_subscribe(mutations: ["createPost","upvotePost","deletePost"])
}
```

Then in your app, you'd set up your `sync` call with:

**Base Query**

The initial `listRecords()` query (no `nextToken) that returns the first page of results. You would presumably order this so that the results are paginated in a way that makes sense for your app user--e.g., most recently-updated first.

**Delta Query**

Basically, this is just the base query with a "last sync" time. It would be re-invoked (assuming the refresh interval hadn't expired) to retrieve the first page of records that had been revised since the last sync.

The pagination for the above queries would then be handled in their respective query callbacks: from within the query callback, you'd issue an `appsync.fetch(ListRecords(nextToken: nextTokenFromPage1Result)`.

While the results of the `listRecords(nextToken:)` and `listRecordsDelta(lastSyncTime:nextToken:)` call would automatically be cached by AppSync, it probably doesn't make sense to attempt to retrieve cached results for anything but page 1.

The local cache would still be populated with individual objects returned by the query, so even if the result pagination changes, the individual records would still be available to provide faster UI updates.

That leaves the question of how you subscribe to state updates.

**Subscription**

In your app, you would pass an `OnDeltaPostSubscription` as the subscription to your `sync` call. The subscription would return individual subscription notifications, one per mutation, that would then be cached to let you maintain local state. The subscription would potentially include events for which you haven't retrieved paginated results (e.g., I've retrieved records 1-100; subscription returns results for record 101), but the subscription handler could reconcile that by either adding the new record to the local cache, or by discarding the subscription result if it's not being used.

So with that in mind, I'd correct and clarify the flow from my comment above to look like:

- App launches for first time
- App executes `listRecords()` query
- Base query returns `nil` from cache
- Base query returns ResultSet1 from service, with NextToken1 for continuation
- AppSync stores RS1/NT1 as cached result for `listRecords()`
- *From within the base query result handler* the app executes `listRecords(NT1)` to get next page of results
- Base query returns ResultSet2 from service, with NextToken2 for continuation
- AppSync stores RS2/NT2 to local cache as cached result for `listRecords(NT1)`
- App quits & resumes
- Delta sync kicks off, starting with BaseQuery of `listRecords()`
- AppSync returns RS1/NT1 from local cache
  - AppSync begins retrieving RS1a/NT1a from service. Meanwhile...
  - ...in delta/base query result handler, recognizing that the source is `cache`, the handler does *not* invoke `listRecords(nextToken:NT1)`, but instead waits for an updated result set from the server
- AppSync stores RS1a/NT1a as updated cached result for `listRecords()`
- *From within the base query result handler* the app executes `listRecords(NT1a)` to get next page of updated/delta results

Does that answer your question, or am I confusing matters even further? :)

**Comment by javiertc** [comment]
Hi @palpatim, thanks for the explanation, really appreciated. I'm going to work on it.

**Comment by airstance** [comment]
Yes, I believe this answered my question.  I now iteratively issue additional fetch calls within the sync completion handler.

One thing to note is that if you save the cancellable object return from the additional fetch calls, be sure to distinguish between cached and server fetches.  They can happen concurrently because the sync call may immediately query the server after the initial cache load.
